### PR TITLE
Node Capacity from Metrics

### PR DESCRIPTION
**Pull request by r-dilip**

### CODE REVIEW

File: source/plugins/ruby/kubelet_utils.rb [hunk 1]
```diff
@@ -17,11 +18,14 @@ def get_node_capacity
...
         memory_capacity = 1.0
 
-        response = CAdvisorMetricsAPIClient.getNodeCapacityFromCAdvisor(winNode: nil)
+        response = CAdvisorMetricsAPIClient.getAllMetricsCAdvisor(winNode: nil)
         if !response.nil? && !response.body.nil?
-          cpu_capacity = JSON.parse(response.body)["num_cores"].nil? ? 1.0 : (JSON.parse(response.body)["num_cores"] * 1000.0)
-          memory_capacity = JSON.parse(response.body)["memory_capacity"].nil? ? 1.0 : JSON.parse(response.body)["memory_capacity"].to_f
-          @log.info "CPU = #{cpu_capacity}mc Memory = #{memory_capacity / 1024 / 1024}MB"
+          all_metrics = response.body.split("\n")
+          cpu_capacity = all_metrics.select{|m| m.start_with?('machine_cpu_cores') }.first.split.last.to_f * 1000
```
Comment (vishiy) [inline source/plugins/ruby/kubelet_utils.rb:24]: did you check the payload that no other metric in it has these 2 metric names as prefix ?

Comment (vishiy) [inline source/plugins/ruby/kubelet_utils.rb:24]: better to check the first element after the spilt is indeed 'machine_cpu_cores' for future proofing, so newer metrics with this prefix doesn't break us.

File: source/plugins/ruby/kubelet_utils.rb [hunk 2]
```diff
@@ -17,11 +18,14 @@ def get_node_capacity
         cpu_capacity = 1.0
         memory_capacity = 1.0
 
-        response = CAdvisorMetricsAPIClient.getNodeCapacityFromCAdvisor(winNode: nil)
+        response = CAdvisorMetricsAPIClient.getAllMetricsCAdvisor(winNode: nil)
```
Comment (ganga1980) [inline source/plugins/ruby/kubelet_utils.rb:21]: these calls can also fail, probably checking for status code and also emitting telemetry is good idea.

File: source/plugins/ruby/kubelet_utils.rb [hunk 3]
```diff
@@ -17,11 +18,14 @@ def get_node_capacity
...
         if !response.nil? && !response.body.nil?
-          cpu_capacity = JSON.parse(response.body)["num_cores"].nil? ? 1.0 : (JSON.parse(response.body)["num_cores"] * 1000.0)
-          memory_capacity = JSON.parse(response.body)["memory_capacity"].nil? ? 1.0 : JSON.parse(response.body)["memory_capacity"].to_f
-          @log.info "CPU = #{cpu_capacity}mc Memory = #{memory_capacity / 1024 / 1024}MB"
+          all_metrics = response.body.split("\n")
+          cpu_capacity = all_metrics.select{|m| m.start_with?('machine_cpu_cores') }.first.split.last.to_f * 1000
+          @log.info "CPU Capacity #{cpu_capacity}"
+          memory_capacity_e = all_metrics.select{|m| m.start_with?('machine_memory_bytes') }.first.split.last
```
Comment (ganga1980) [inline source/plugins/ruby/kubelet_utils.rb:26]: as vishwa commented in other comment, having strict check on metric name will ensure we are picking the right metric.

File: source/plugins/ruby/kubelet_utils.rb (hunk 1 above)
Comment (r-dilip) [reply source/plugins/ruby/kubelet_utils.rb:24]: doing an exact comparison

Comment (r-dilip) [reply source/plugins/ruby/kubelet_utils.rb:24]: Done

File: source/plugins/ruby/kubelet_utils.rb (hunk 2 above)
Comment (r-dilip) [reply source/plugins/ruby/kubelet_utils.rb:21]: we are already emitting telemetry on failure here. (L33)

Comment (ganga1980) [reply source/plugins/ruby/kubelet_utils.rb:21]: > we are already emitting telemetry on failure here. (L33)

> @r-dilip, what I am saying,  extract metrics only when response.code == 200 and if its not 200, just emit the telemetry.  this we can avoid sending un-necessarily full call stack for all non-200 responses.

**Pull request merged**